File: frontendmasters/rust-for-typescript-devs/typescript/src/fissures.ts
import { readFileSync } from "fs";
import { join } from "path";

function getInput(): string {
    return readFileSync(join(__dirname, "..", "..", "fissures.txt")).toString();
  }

type Point = {
    x: number,
    y: number
}

type Line = {
    p1: Point,
    p2: Point
}

function isHOrV({p1, p2}: Line) {
    return p1.x === p2.x || p1.y === p2.y
}

function parsePoint(p: string) {
    const [x,y] = p.split(',');
    return {
        x: +x,
        y: +y
    }
}

function parseLine(p: string) {
    const [p1, p2] = p.split(' ->');
    return {
        p1: parsePoint(p1),
        p2: parsePoint(p2)
    }
}

const lines = getInput().split("\n")
    .map(x => parseLine(x))
    .filter(isHOrV);

console.log(lines)

File: frontendmasters/rust-for-typescript-devs/rust/src/bin/fissures.rs
use anyhow::{anyhow, Result, Context};
use std::str::FromStr;

fn get_input(file_path: &str) -> String {
    return std::fs::read_to_string(file_path).expect("Should have been able to read the file");
}

#[derive(Debug)]
struct Point {
    x: i32,
    y: i32,
}

#[derive(Debug)]
struct Line {
    p1: Point,
    p2: Point,
}

impl FromStr for Point {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self> {
        let result = s.split_once(",");
        if result.is_none() {
            return Err(anyhow!("expected a point to contain a comma"));
        }
        let (x, y) = result.unwrap();
        let x: i32 = str::parse(x)?;
        let y: i32 = str::parse(y)?;

        return Ok(Point { x, y });
    }
}

impl FromStr for Line {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self> {
        let result = s.split_once(" -> ");

        if result.is_none() {
            return Err(anyhow!("expected a line to contain ->"));
        }
        let (p1, p2) = result.unwrap();
        let p1 = str::parse(p1)?;
        let p2 = str::parse(p2)?;

        return Ok(Line { p1, p2 });
    }
}

impl Line {
    fn is_horv(&self) -> bool {
        return self.p1.x == self.p2.x || self.p1.y == self.p2.y;
    }
}

fn main() {
    let data = get_input("../../../fissures.txt");

    let lines = data
        .lines()
        .flat_map(str::parse)
        .filter(|x: &Line| x.is_horv())
        .collect::<Vec<Line>>();

    println!("{:?}", lines);
}
